fix pmi2 process map parsing in ess/pmi reusing the correct one in grpomm/pmi

plesn opened the pull request:
This fixes launching of openmpi-1.8 with mxm or sm in cases of
unpopulated nodes like this :

    $ srun  -N2 -n7  --cpu_bind=core -m block --mpi=pmi2 -K ./IMB-MPI1 Pingpong
    [btp19:24155] Error: mtl_mxm.c:310 - ompi_mtl_mxm_module_init() Unable to obtain local node rank
    [btp19:24157] Error: mtl_mxm.c:310 - ompi_mtl_mxm_module_init() Unable to obtain local node rank
    --------------------------------------------------------------------------
    WARNING: Missing locality information required for sm initialization.
    Continuing without shared memory support.
    --------------------------------------------------------------------------
    [...]

Here mxm and sm need the local/node rank, whose parsing in ess/pmi is
buggy when Slurm gives process mapping vectors with an offset like
`(1,X,X)`. There is already a function that correctly reads Slurm's
process mapping in grpcomm/pmi/pmi2_pmap_parser.c.

I propose to fix the bug by reusing this pmap parsing function, and to
be a little bit prettier in the 2nd commit I move this common function
to opal/mca/common/pmi.

If this seems okay to you, the pmi process mapping query code could be
factorised even further in a `mca_common_pmapping()` function.

mellanox-github commented:
- Coverity found 1465 errors for all
- Coverity found 23 errors for oshmem
- Coverity found 7 errors for fca
- Coverity report at http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-release-v1.8-pr//327/Coverity_Report <br><br>:octocat: _Sent from [GH](http://nodegh.io)._

mellanox-github commented:
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/job/gh-ompi-release-v1.8-pr/327/

rhc54 commented:
hmmm...we've had problems in the past with issues caused by changes in slurm versions and/or inconsistencies between their PMI-1 and PMI-2 implementations. This code has been working for a long time - I see you removed the current code to replace it with the PMI-2 parser. So does that mean that PMI-1 is no longer supported? Do we have issues with non-Slurm PMI's?

Any issues with earlier versions of Slurm? I guess that begs the question - what version of Slurm were you using?

plesn commented:
This bug is PMI-2 only : I modified the part inside `#if
WANT_PMI2_SUPPORT` and did not remove any PMI-1 code.

With the mapping `(vector,(0,1,6),(1,1,1))` returned by Slurm the
parsing loop in `rte_init()` incorectly parses local ranks;

Currently there are 2 places where the PMI-2 process mapping is
independently parsed:
- in ess/pmi in rte_init() : this one is incorrect.
  It is used to compute the local rank / node rank / local leader
- in grpcomm/pmi/pmi2_pmap_parser.c : this one seems correct and has
  tests, so I propose to reuse this one.
  It is use to compute local ranks in the modex

PS: I renamed the parsing function to put in common/pmi, but I forgot
to rename it in the standalone tests

mellanox-github commented:
- Coverity found 1465 errors for all
- Coverity found 23 errors for oshmem
- Coverity found 7 errors for fca
- Coverity report at http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-release-v1.8-pr//329/Coverity_Report <br><br>:octocat: _Sent from [GH](http://nodegh.io)._

mellanox-github commented:
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/job/gh-ompi-release-v1.8-pr/329/

plesn commented:
Added `mca_common_pmi_local_ranks()` to avoid duplication between
ess/pmi and grpcom/pmi of code that discovers local ranks in
PMI1/2. Maybe you will find it clearer this way.

Still the bug is corrected by just using the very first commit.

rhc54 commented:
Thanks! Have you signed a contributor agreement, or belong to one of our member groups, perchance?

plesn commented:
It didn't sign anything like this. My organisation is Bull, but I can't find if it is in member groups.
Edit: Bull is a "contributor organisation"

mellanox-github commented:
- Coverity found 1466 errors for all
- Coverity found 23 errors for oshmem
- Coverity found 7 errors for fca
- Coverity report at http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-release-v1.8-pr//330/Coverity_Report <br><br>:octocat: _Sent from [GH](http://nodegh.io)._

mellanox-github commented:
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/job/gh-ompi-release-v1.8-pr/330/

rhc54 commented:
Yep, Bull is fine - they've been a member for years. Thanks again!

rhc54 commented:
Looks okay to me - has no bearing on master as the code there is totally different.

rhc54 commented:
@plesn Could you please create a separate PR with the same changes against the v1.10 branch?

plesn commented:
@rhc54  my last commit is erroneous in PMI1: I realise it was still compiling in pmi2 when I thought I was selecting the pmi1 version!

So I must add the correction in a separate PR from what I see, and the equivalent rebased on v1.10
